install() function not self-retriggering properly

[This is with v5.2]

After manually running the install() function once, it successfully triggers the startSync() function to run a few moments later, but the trigger to re-run the install() function again after howFrequent minutes happens only once, and that instance does not trigger startSync() or install() again. The syncing happens properly the one time (good logs, no errors, entries properly synced). The install() function does not output any logs.

i.e. in the logs I see, i sequence, install() run manually (by Editor), then startSync() run by trigger, then howFrequent minutes later install() run by trigger. And then nothing else after that.

Every few days I come back and manually run the install() function again, and the same pattern of executions as described above happens again. The expected continuous re-triggering of install() is not happening.

Enabled logging of the exception of the catch{} block as you suggested, which resulted in nothing being logged, so I don't think an exception was being thrown. So I then added additional logging before and after each trigger add as follows:

function install(){
  try{
    //Delete any already existing triggers so we don't create excessive triggers
    deleteAllTriggers();
    
    if (howFrequent < 1){
      throw "[ERROR] \"howFrequent\" must be greater than 0.";
    }
    else{
      Logger.log("About to set install() trigger for now+25 minutes");
      ScriptApp.newTrigger("install").timeBased().after(howFrequent * 60 * 1000).create();//Schedule next Execution
      Logger.log("Done set install() trigger for now+25 minutes");
      Logger.log("About to set startSync() trigger for now+1 second");
      ScriptApp.newTrigger("startSync").timeBased().after(1000).create();//Start the sync routine
      Logger.log("Done set startSync() trigger for now+1 second");
    }
  }catch(e){
    Logger.log(e);
    //#install();//Retry on error
  }
}

The same pattern of executions resulted, before no further executions:

ICS-Syncer | install | Trigger | Jun 22, 2020, 3:34:45 PM | 32.393 s | Completed
ICS-Syncer | startSync | Trigger | Jun 22, 2020, 3:10:08 PM | 2.867 s | Completed
ICS-Syncer | install | Editor | Jun 22, 2020, 3:09:08 PM | 1.661 s | Completed


Both instances of install() logged the same info, like this:

Jun 22, 2020, 3:34:46 PM Info About to set install() trigger for now+25 minutes
Jun 22, 2020, 3:35:07 PM Info Done set install() trigger for now+25 minutes
Jun 22, 2020, 3:35:07 PM Info About to set startSync() trigger for now+1 second
Jun 22, 2020, 3:35:17 PM Info Done set startSync() trigger for now+1 second

Notice that sometimes the calls to create a new trigger take an exceptionally long time (21 seconds, and 10 seconds, respectively in this instance). I don't *think* that's relevant, but you may know better.

So even though the second instance of install() (triggered by the first manual run) seems to have successfully added two new time triggers (because there was no exception, and the "Done" logging happens), they don't ever actually fire.

Hi, 

I had a similar problem and tried a few things.
To resolve this, I disabled the trigger for install (why do you have to reinstall every hour?) - 

And I changed the trigger generator for startSync to a different method: 
https://developers.google.com/apps-script/reference/script/clock-trigger-builder#everyminutesn

Here's what I have now, startSync seems to run fine at the set interval: 

``` function install(){
  try{
    //Delete any already existing triggers so we don't create excessive triggers
    deleteAllTriggers();
    
    if (howFrequent < 1){
      throw "[ERROR] \"howFrequent\" must be greater than 0.";
    }
    else{
      //ScriptApp.newTrigger("install").timeBased().everyHours(24).create();//Schedule install interval
      ScriptApp.newTrigger("startSync").timeBased().everyMinutes(15).create();//schedule sync interval
      startSync(); //immediate sync
      
    }
  }catch(e){
    install();//Retry on error
  }
}
```
I will now try to set the install trigger to a longer interval, prob every week:
https://developers.google.com/apps-script/reference/script/clock-trigger-builder#everyweeksn
Although I am still not sure why that is actually necessary?


Either way, I hope that helps,
U

New install function: 

```
function install(){
  try{
    //Delete any already existing triggers so we don't create excessive triggers
    deleteAllTriggers();
    
    if (howFrequent < 1){
      throw "[ERROR] \"howFrequent\" must be greater than 0.";
    }
    else{
      ScriptApp.newTrigger("install").timeBased().everyWeeks(1).create();//Schedule next Execution
      ScriptApp.newTrigger("startSync").timeBased().everyMinutes(howFrequent).create();//Start the sync routine
      startSync();
      
    }
  }catch(e){
    install();//Retry on error
  }
}
```

On this topic, I have been having the same issue as others have mentioned of this thread (install script will randomly stop re-triggering after a while and my calendar ends up stale) so I modified the script to use the `everyMinutes` trigger setting as others mentioned here. 

Additionally, to make the code more generally usable, I made a helper function to still use the `howFrequent` value as the user's intended amount of time but round it to the closest valid time setting usable for `everyMinutes` (1, 5 10, 15, 30). Hopefully this is a good middle ground to get the reliability of everyMinutes and also still capture the rough intent of how long a user wants to have the script wait between re-syncs.

In Code.gs, I modified `install`:
```
function install(){
  try{
    //Delete any already existing triggers so we don't create excessive triggers
    deleteAllTriggers();
    
    if (howFrequent < 1){
      throw "[ERROR] \"howFrequent\" must be greater than 0.";
    }
    else{
      ScriptApp.newTrigger("startSync").timeBased().everyMinutes(getValidTriggerFrequency(howFrequent)).create(); //Schedule sync routine to explicitly repeat
      ScriptApp.newTrigger("startSync").timeBased().after(1000).create();//Start the first sync routine immediately
    }
  }catch(e){
    install();//Retry on error
  }
}
```

and in Helpers.gs, I added `getValidTriggerFrequency`: 

```
function getValidTriggerFrequency(origFrequency) {
  if (origFrequency == null) 
    return 1; // if param is empty, just return 1, the lowest valid value.
  
  var adjFrequency = Math.round(origFrequency/5) * 5; // Set the number to be the closest divisible-by-5
  adjFrequency = Math.max(adjFrequency, 1); // Make sure the number is at least 1 (0 is not valid for the trigger)
  adjFrequency = Math.min(adjFrequency, 15); // Make sure the number is at most 15 (will check for the 30 value below)
  
  if((adjFrequency == 15) && (Math.abs(origFrequency-30) < Math.abs(origFrequency-15)))
    adjFrequency = 30; // If we adjusted to 15, but the original number is actually closer to 30, set it to 30 instead
  
  Logger.log("Intended frequency = "+origFrequency+", Adjusted frequency = "+adjFrequency);
  return adjFrequency;
}
```

I know that's probably not the most elegant way to do the calculation, but it works well enough for me.

Can one of you please test if the following install routine is more stable:
```
function install(){
  //Delete any already existing triggers so we don't create excessive triggers
  deleteAllTriggers();
  
  if (howFrequent < 1)
    throw "[ERROR] \"howFrequent\" must be greater than 0.";
  
  var result = callWithBackoff(function(){
    ScriptApp.newTrigger("install").timeBased().after(howFrequent * 60 * 1000).create();//Schedule next Execution
    
    var triggers = ScriptApp.getProjectTriggers().filter(function(trigger){
      return trigger.getHandlerFunction() == "install";
    });
    
    if (triggers.length < 1)
      throw "[ERROR] Failed to create trigger for next execution.";
    
    return true;
  }, 10);
  
  if (result){
    ScriptApp.newTrigger("startSync").timeBased().after(1000).create();//Start the sync routine
  }
  else{
    throw "[ERROR] Installing the script failed."
  }
}
```

You want me to replace the following with what you wrote above?:
function install(){
  try{
    //Delete any already existing triggers so we don't create excessive triggers
    deleteAllTriggers();
    
    if (howFrequent < 1){
      throw "[ERROR] \"howFrequent\" must be greater than 0.";
    }
    else{
      ScriptApp.newTrigger("install").timeBased().after(howFrequent * 60 * 1000).create();//Schedule next Execution
      ScriptApp.newTrigger("startSync").timeBased().after(1000).create();//Start the sync routine
    }
  }catch(e){
    install();//Retry on error
  }
}

--- a/Code.gs
+++ b/Code.gs
@@ -90,13 +90,8 @@ function install(){
     //Delete any already existing triggers so we don't create excessive triggers
     deleteAllTriggers();
     
-    if (howFrequent < 1){
-      throw "[ERROR] \"howFrequent\" must be greater than 0.";
-    }
-    else{
-      ScriptApp.newTrigger("startSync").timeBased().everyMinutes(getValidTriggerFrequency(howFrequent)).create(); //Schedule sync routine to explicitly repeat
-      ScriptApp.newTrigger("startSync").timeBased().after(1000).create();//Start the sync routine
-    }
+    ScriptApp.newTrigger("startSync").timeBased().everyMinutes(getValidTriggerFrequency(howFrequent)).create(); //Schedule sync routine to explicitly repeat
+    ScriptApp.newTrigger("startSync").timeBased().after(1000).create();//Start the sync routine
   }catch(e){
     install();//Retry on error
   }

--- a/Code.gs
+++ b/Code.gs
@@ -94,7 +94,7 @@ function install(){
       throw "[ERROR] \"howFrequent\" must be greater than 0.";
     }
     else{
-      ScriptApp.newTrigger("install").timeBased().after(howFrequent * 60 * 1000).create();//Schedule next Execution
+      ScriptApp.newTrigger("startSync").timeBased().everyMinutes(getValidTriggerFrequency(howFrequent)).create(); //Schedule sync routine to explicitly repeat
       ScriptApp.newTrigger("startSync").timeBased().after(1000).create();//Start the sync routine
     }
   }catch(e){

--- a/Helpers.gs
+++ b/Helpers.gs
@@ -1,3 +1,28 @@
+/**
+ * Takes an intended frequency in minutes and adjusts it to be the closest 
+ * acceptable value to use Google "everyMinutes" trigger setting (i.e. one of
+ * the following values: 1, 5, 10, 15, 30).
+ *
+ * @param {integer} The manually set frequency that the user intends to set
+ * @return {integer} The closest valid value to the intended frequency setting 
+ */
+function getValidTriggerFrequency(origFrequency) {
+  if (origFrequency == null) 
+    return 1; // if param is empty, just return 1, the lowest valid value.
+
+  
+  var adjFrequency = Math.round(origFrequency/5) * 5; // Set the number to be the closest divisible-by-5
+  adjFrequency = Math.max(adjFrequency, 1); // Make sure the number is at least 1 (0 is not valid for the trigger)
+  adjFrequency = Math.min(adjFrequency, 15); // Make sure the number is at most 15 (will check for the 30 value below)
+  
+  
+  if((adjFrequency == 15) && (Math.abs(origFrequency-30) < Math.abs(origFrequency-15)))
+    adjFrequency = 30; // If we adjusted to 15, but the original number is actually closer to 30, set it to 30 instead
+  
+  Logger.log("Intended frequency = "+origFrequency+", Adjusted frequency = "+adjFrequency);
+  return adjFrequency;
+}
+
 String.prototype.includes = function(phrase){ 
   return this.indexOf(phrase) > -1;
 }

--- a/Helpers.gs
+++ b/Helpers.gs
@@ -9,7 +9,7 @@
 function getValidTriggerFrequency(origFrequency) {
   if (origFrequency == null) {
     Logger.log("No intended frequency specified. Defaulting to 15 minutes.");
-    return 15; // if param is empty, just return 15, the lowest valid value.
+    return 15; // if param is empty, just return the default value of 15.
   }
 
   

--- a/Helpers.gs
+++ b/Helpers.gs
@@ -8,8 +8,8 @@
  */
 function getValidTriggerFrequency(origFrequency) {
   if (origFrequency == null) {
-    Logger.log("No frequency specified. Defaulting to 15 minutes.");
-    return 15; // if param is empty, just return 1, the lowest valid value.
+    Logger.log("No intended frequency specified. Defaulting to 15 minutes.");
+    return 15; // if param is empty, just return 15, the lowest valid value.
   }
 
   

--- a/Helpers.gs
+++ b/Helpers.gs
@@ -3,13 +3,13 @@
  * acceptable value to use Google "everyMinutes" trigger setting (i.e. one of
  * the following values: 1, 5, 10, 15, 30).
  *
- * @param {integer} The manually set frequency that the user intends to set. If blank, 15 is assumeed.
- * @return {integer} The closest valid value to the intended frequency setting 
+ * @param {?integer} The manually set frequency that the user intends to set.
+ * @return {integer} The closest valid value to the intended frequency setting. Defaulting to 15 if no valid input is provided.
  */
 function getValidTriggerFrequency(origFrequency) {
-  if (origFrequency == null) {
-    Logger.log("No intended frequency specified. Defaulting to 15 minutes.");
-    return 15; // if param is empty, just return the default value of 15.
+  if (!origFrequency > 0) {
+    Logger.log("No valid frequency specified. Defaulting to 15 minutes.");
+    return 15;
   }
 
   

--- a/Helpers.gs
+++ b/Helpers.gs
@@ -3,12 +3,14 @@
  * acceptable value to use Google "everyMinutes" trigger setting (i.e. one of
  * the following values: 1, 5, 10, 15, 30).
  *
- * @param {integer} The manually set frequency that the user intends to set
+ * @param {integer} The manually set frequency that the user intends to set. If blank, 15 is assumeed.
  * @return {integer} The closest valid value to the intended frequency setting 
  */
 function getValidTriggerFrequency(origFrequency) {
-  if (origFrequency == null) 
-    return 1; // if param is empty, just return 1, the lowest valid value.
+  if (origFrequency == null) {
+    Logger.log("No frequency specified. Defaulting to 15 minutes.");
+    return 15; // if param is empty, just return 1, the lowest valid value.
+  }
 
   
   var adjFrequency = Math.round(origFrequency/5) * 5; // Set the number to be the closest divisible-by-5

--- a/Helpers.gs
+++ b/Helpers.gs
@@ -11,12 +11,10 @@ function getValidTriggerFrequency(origFrequency) {
     Logger.log("No valid frequency specified. Defaulting to 15 minutes.");
     return 15;
   }
-
   
   var adjFrequency = Math.round(origFrequency/5) * 5; // Set the number to be the closest divisible-by-5
   adjFrequency = Math.max(adjFrequency, 1); // Make sure the number is at least 1 (0 is not valid for the trigger)
   adjFrequency = Math.min(adjFrequency, 15); // Make sure the number is at most 15 (will check for the 30 value below)
-  
   
   if((adjFrequency == 15) && (Math.abs(origFrequency-30) < Math.abs(origFrequency-15)))
     adjFrequency = 30; // If we adjusted to 15, but the original number is actually closer to 30, set it to 30 instead

